File: Shared/Engine/MaxEditDraft.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import Foundation

/// A manual edit starts from exact hand records, never their shared fallback.
/// Opening the editor, converting display units, or leaving one hand alone must
/// not create another benchmark. Explicit changes or selected retests become
/// append-only saves; an unchanged retest never quietly includes the other hand.
struct MaxEditDraft: Equatable, Sendable {
    struct Change: Equatable, Sendable {
        let side: Side
        let kg: Double
    }

    private var originals: [Side: Double]
    private(set) var repeatedTests: Set<Side> = []
    var leftKg: Double
    var rightKg: Double

    init(leftKg: Double? = nil, rightKg: Double? = nil) {
        var originals: [Side: Double] = [:]
        if let leftKg, leftKg.isFinite, leftKg > 0 { originals[.left] = leftKg }
        if let rightKg, rightKg.isFinite, rightKg > 0 { originals[.right] = rightKg }
        self.originals = originals
        self.leftKg = originals[.left] ?? 0
        self.rightKg = originals[.right] ?? 0
    }

    func originalKg(for side: Side) -> Double? { originals[side] }

    func recordsAnotherTest(for side: Side) -> Bool { repeatedTests.contains(side) }

    mutating func setRecordsAnotherTest(_ enabled: Bool, for side: Side) {
        guard side != .both else { return }
        if enabled {
            guard originals[side] != nil else { return }
            repeatedTests.insert(side)
        } else {
            repeatedTests.remove(side)
        }
    }

    /// History can change while this editor remains on its navigation stack.
    /// Follow those changes only for untouched fields; a typed draft belongs to the
    /// person editing it. Compare any retained edits against the latest exact max.
    mutating func rebase(leftKg: Double?, rightKg: Double?) {
        let leftWasEdited = self.leftKg != (originals[.left] ?? 0) || repeatedTests.contains(.left)
        let rightWasEdited = self.rightKg != (originals[.right] ?? 0) || repeatedTests.contains(.right)
        let current = MaxEditDraft(leftKg: leftKg, rightKg: rightKg)
        originals = current.originals
        if !leftWasEdited { self.leftKg = current.leftKg }
        if !rightWasEdited { self.rightKg = current.rightKg }
    }

    private func value(for side: Side) -> Double {
        side == .left ? leftKg : rightKg
    }

    /// Zero cannot mean delete: removing a current record exposes its predecessor.
    /// Reject the entire edit if a changed field is invalid so Save cannot silently
    /// ignore that hand while committing the other one.
    var hasInvalidChanges: Bool {
        [Side.left, .right].contains { side in
            let kg = value(for: side)
            return (kg != (originals[side] ?? 0) || repeatedTests.contains(side))
                && (!kg.isFinite || kg <= 0)
        }
    }

    var changes: [Change] {
        [Side.left, .right].compactMap { side in
            let kg = value(for: side)
            guard kg.isFinite, kg > 0,
                  kg != (originals[side] ?? 0) || repeatedTests.contains(side) else { return nil }
            return Change(side: side, kg: kg)
        }
    }

    var canSave: Bool { !hasInvalidChanges && !changes.isEmpty }
}

File: android/engine/src/main/kotlin/run/nuri/getagrip/engine/MaxEditDraft.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.engine

/** Unsaved hand values. Existing shared fallbacks never become invented hand records. */
class MaxEditDraft(leftKg: Double? = null, rightKg: Double? = null) {
    data class Change(val side: Side, val kg: Double)

    private var originals: Map<Side, Double> = buildMap {
        leftKg?.takeIf { it.isFinite() && it > 0 }?.let { put(Side.left, it) }
        rightKg?.takeIf { it.isFinite() && it > 0 }?.let { put(Side.right, it) }
    }
    var repeatedTests: Set<Side> = emptySet()
        private set
    var leftKg: Double = originals[Side.left] ?: 0.0
    var rightKg: Double = originals[Side.right] ?: 0.0

    fun originalKg(side: Side): Double? = originals[side]
    fun recordsAnotherTest(side: Side): Boolean = side in repeatedTests

    fun setRecordsAnotherTest(enabled: Boolean, side: Side) {
        if (side == Side.both) return
        if (enabled && originals[side] != null) repeatedTests = repeatedTests + side
        else if (!enabled) repeatedTests = repeatedTests - side
    }

    /** Follow history changes for untouched fields, preserving edited or explicitly retested values. */
    fun rebase(leftKg: Double?, rightKg: Double?) {
        val leftWasEdited = this.leftKg != (originals[Side.left] ?: 0.0) || Side.left in repeatedTests
        val rightWasEdited = this.rightKg != (originals[Side.right] ?: 0.0) || Side.right in repeatedTests
        val current = MaxEditDraft(leftKg, rightKg)
        originals = current.originals
        if (!leftWasEdited) this.leftKg = current.leftKg
        if (!rightWasEdited) this.rightKg = current.rightKg
    }

    private fun value(side: Side): Double = if (side == Side.left) leftKg else rightKg

    val hasInvalidChanges: Boolean get() = listOf(Side.left, Side.right).any { side ->
        val kg = value(side)
        (kg != (originals[side] ?: 0.0) || side in repeatedTests) && (!kg.isFinite() || kg <= 0)
    }

    val changes: List<Change> get() = listOf(Side.left, Side.right).mapNotNull { side ->
        val kg = value(side)
        if (kg.isFinite() && kg > 0 && (kg != (originals[side] ?: 0.0) || side in repeatedTests))
            Change(side, kg) else null
    }

    val canSave: Boolean get() = !hasInvalidChanges && changes.isNotEmpty()

    /** Value-style use with Compose mutableStateOf; the engine stays platform independent. */
    fun copy(): MaxEditDraft = MaxEditDraft(originals[Side.left], originals[Side.right]).also {
        it.leftKg = leftKg
        it.rightKg = rightKg
        it.repeatedTests = repeatedTests.toSet()
    }
}
